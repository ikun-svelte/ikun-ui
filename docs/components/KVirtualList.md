---
title: KVirtualList
lang: en-US
---

# KVirtualList

A simple virtual list, useful when rendering large amounts of data.

> The code implementation of `KVirtualList` comes from [svelte-virtual-scroll-list <span class="i-carbon-link text-12px" />](https://github.com/v1ack/svelte-virtual-scroll-list)

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/virtual-list
```

```bash [yarn]
yarn add @ikun-ui/virtual-list
```

```bash [npm]
npm install @ikun-ui/virtual-list
```

:::

## Basic usage

Without further ado, please take a look at the demo

<demo src="virtual-list/basic.svelte"  github='VirtualList'></demo>

## Different heights

You can set different heights for `item` according to your needs.

<demo src="virtual-list/height.svelte"  github='VirtualList'></demo>

## Horizontal virtual lists

`isHorizontal` can control the display of horizontal virtual lists.

<demo src="virtual-list/horizontal.svelte"  github='VirtualList'></demo>

## Custom rendering header and footer

Slots allow you to add content at the top and bottom of the list.

<demo src="virtual-list/slot.svelte"  github='VirtualList'></demo>

## Scroll to bottom

KVirtualList provides some APIs, such as `scrollToBottom`.

<demo src="virtual-list/bottom.svelte"  github='VirtualList'></demo>

## VirtualList Props

| Name            | Type                         | Default | Description                                                        |
| --------------- | ---------------------------- | ------- | ------------------------------------------------------------------ |
| data            | `Array<Record<string, any>>` | `-`     | Source for list.                                                   |
| key             | `string`                     | `'id'`  | Unique key for getting data from `data`                            |
| keeps           | `number`                     | `30`    | Count of rendered items                                            |
| estimateSize    | `number`                     | `50`    | Estimate size of each item, needs for smooth scrollbar.            |
| isHorizontal    | `boolean`                    | `false` | Scroll direction.                                                  |
| start           | `number`                     | `0`     | Scroll position start index                                        |
| offset          | `number`                     | `0`     | Scroll position offset                                             |
| topThreshold    | `number`                     | `0`     | The threshold to emit `top` event, attention to multiple calls.    |
| bottomThreshold | `number`                     | `0`     | The threshold to emit `bottom` event, attention to multiple calls. |
| cls             | `string`                     | `''`    | Additional class for component                                     |
| attrs           | `Record<string, string>`     | `{}`    | Additional attributes                                              |

## VirtualList Events

| Name   | Description                                  | Type                                                 |
| ------ | -------------------------------------------- | ---------------------------------------------------- |
| scroll | trigger when scroll.                         | `(data: {event: CustomEvent, range: VRange})=> void` |
| top    | Fires when the scroll bar reaches the top    | `()=> void`                                          |
| bottom | Fires when the scroll bar reaches the bottom | `()=> void`                                          |

```typescript
declare interface VRange {
	start: number;
	end: number;
	padFront: number;
	padBehind: number;
}
```

## VirtualList Slots

| Name    | Description                   |
| ------- | ----------------------------- |
| default | Customize item content        |
| footer  | Customize footer item content |
| header  | Customize header item content |

## VirtualList Api

| method         | arguments              | description                                                |
| -------------- | ---------------------- | ---------------------------------------------------------- |
| scrollToBottom | `none`                 | Scroll list to bottom                                      |
| scrollToIndex  | `index: number`        | Set scroll position to a designated index                  |
| scrollToOffset | `offset: number`       | Set scroll position to a designated offset                 |
| getSize        | `id: typeof props.key` | Get the designated item size                               |
| getSizes       | `none`                 | Get the total number of stored (rendered) items            |
| getOffset      | `none`                 | Get current scroll offset                                  |
| getClientSize  | `none`                 | Get wrapper element client viewport size (width or height) |
| getScrollSize  | `none`                 | Get all scroll size (scrollHeight or scrollWidth)          |
