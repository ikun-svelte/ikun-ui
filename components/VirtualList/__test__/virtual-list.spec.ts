import { afterEach, expect, test, describe, beforeEach, vi } from 'vitest';
import KVirtualList from '../src';
let host;
const dataList: any = [];
for (let i = 0; i < 1000; i++) {
	dataList.push({ id: i, label: `this is item ${i}` });
}
const initHost = () => {
	host = globalThis.document.createElement('div');
	host.setAttribute('id', 'host');
	host.style.height = '50px';
	host.style.overflow = 'auto';
	globalThis.document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.restoreAllMocks();
});

describe('Test: KVirtualList', () => {
	vi.mock('svelte', async () => {
		const actual = (await vi.importActual('svelte')) as object;
		return {
			...actual,
			// @ts-ignore
			onMount: (await import('svelte/internal')).onMount,
			// @ts-ignore
			onDestroy: (await import('svelte/internal')).onDestroy,
			// @ts-ignore
			afterUpdate: (await import('svelte/internal')).afterUpdate
		};
	});

	test('props: cls', async () => {
		const instance = new KVirtualList({
			target: host,
			props: {
				data: dataList,
				cls: 'k-virtual-list--test'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-virtual-list--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: keeps', async () => {
		const instance = new KVirtualList({
			target: host,
			props: {
				data: dataList,
				keeps: 10
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelectorAll('.k-virtual-list--item').length === 10).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: isHorizontal', async () => {
		const instance = new KVirtualList({
			target: host,
			props: {
				data: dataList,
				isHorizontal: true
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelectorAll('.k-virtual-list--item').length === 30).toBeTruthy();
		expect(host!.innerHTML.includes('k-virtual-list--base')).toBeTruthy();
		expect(host!.innerHTML.includes('k-virtual-list--wrapper')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
