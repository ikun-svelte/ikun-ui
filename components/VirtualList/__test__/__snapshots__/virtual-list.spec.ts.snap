// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KVirtualList > props: cls 1`] = `"<div class=\\"k-virtual-list k-scrollbar k-virtual-list--test\\" style=\\"overflow-y: auto; height: inherit; --k-scrollbar--thumb__radius: 20px; --k-scrollbar--thumb: var(--ikun-light-800); --k-scrollbar__h: 6px; --k-scrollbar__w: 6px; --k-scrollbar--track__radius: 20px; --k-scrollbar--track: transparent;\\"> <div class=\\"\\" style=\\"padding: 0px 0px 9700px;\\"><div class=\\"k-virtual-list--item\\" data-kv-key=\\"0\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"1\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"2\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"3\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"4\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"5\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"6\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"7\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"8\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"9\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"10\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"11\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"12\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"13\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"14\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"15\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"16\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"17\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"18\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"19\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"20\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"21\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"22\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"23\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"24\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"25\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"26\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"27\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"28\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"29\\"> </div></div>  <div class=\\"shepherd\\" style=\\"width: 100%; height: 0px;\\"></div></div>"`;

exports[`Test: KVirtualList > props: isHorizontal 1`] = `"<div class=\\"k-virtual-list k-virtual-list--base k-scrollbar\\" style=\\"overflow-y: auto; height: inherit; --k-scrollbar--thumb__radius: 20px; --k-scrollbar--thumb: var(--ikun-light-800); --k-scrollbar__h: 6px; --k-scrollbar__w: 6px; --k-scrollbar--track__radius: 20px; --k-scrollbar--track: transparent;\\"> <div class=\\"k-virtual-list--wrapper\\" style=\\"padding: 0px 9700px 0px 0px;\\"><div class=\\"k-virtual-list--item\\" data-kv-key=\\"0\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"1\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"2\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"3\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"4\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"5\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"6\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"7\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"8\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"9\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"10\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"11\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"12\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"13\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"14\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"15\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"16\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"17\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"18\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"19\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"20\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"21\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"22\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"23\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"24\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"25\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"26\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"27\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"28\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"29\\"> </div></div>  <div class=\\"shepherd\\" style=\\"width: 0px; height: 100%;\\"></div></div>"`;

exports[`Test: KVirtualList > props: keeps 1`] = `"<div class=\\"k-virtual-list k-scrollbar\\" style=\\"overflow-y: auto; height: inherit; --k-scrollbar--thumb__radius: 20px; --k-scrollbar--thumb: var(--ikun-light-800); --k-scrollbar__h: 6px; --k-scrollbar__w: 6px; --k-scrollbar--track__radius: 20px; --k-scrollbar--track: transparent;\\"> <div class=\\"\\" style=\\"padding: 0px 0px 9900px;\\"><div class=\\"k-virtual-list--item\\" data-kv-key=\\"0\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"1\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"2\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"3\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"4\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"5\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"6\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"7\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"8\\"> </div><div class=\\"k-virtual-list--item\\" data-kv-key=\\"9\\"> </div></div>  <div class=\\"shepherd\\" style=\\"width: 100%; height: 0px;\\"></div></div>"`;
